client.py

```diff
@@ -161,97 +200,92 @@ def close(self):
                 self._sock.connect(serv)
-                self._sock.setblocking(False)
-                # Start reading before server can send: can't send until it
-                # gets ID.
-                loop.create_task(_reader)
-                # Server reads ID immediately, but a brief pause is probably wise.
-                await asyncio.sleep_ms(50)
-                # No need for lock yet.
-                await self._send(self._my_id)  # Can throw OSError
-            except OSError:
-                if init:
-                    await self.bad_server()
-            else:
-                # Improved cancellation code contributed by Kevin Köck
-                # Note _writer pauses before 1st tx
-                _writer = self._writer()
-                loop.create_task(_writer)
+            except OSError as e:
+                if e.args[0] in (errno.ECONNABORTED, errno.ECONNRESET, errno.ECONNREFUSED):
+                    if init:
+                        await self.bad_server()
+                    continue  # temporary server outage?
+            self._sock.setblocking(False)
+            # Start reading before server can send: can't send until it
+            # gets ID.
+            loop.create_task(_reader)
+            # Server reads ID immediately, but a brief pause is probably wise.
+            await asyncio.sleep_ms(50)
+            # last_mid = self._recv_mid
+            self._ack_mid = -1
+            preheader = bytearray(5)
+            preheader[0] = 0x2C  # mid but in this case protocol identifier but will receive an ACK with mid 0x2C
+            preheader[1] = 0  # header length
+            preheader[2] = len(self._my_id) & 0xFF
+            preheader[3] = (len(self._my_id) >> 8) & 0xFF  # allows for 65535 message length
+            preheader[4] = 0xFF  # clean connection, shows if device has been reset or just a wifi outage
+            preheader = ubinascii.hexlify(preheader)
+            ok = await self._write(preheader, None, self._my_id, True, 0x2C, init=True)
+            print("Got login", ok)
+            if ok:
+                self._ok = True  # only set self._ok after sending the id was successful.
+                # self._ack_mid = last_mid  # restore state before sending id to prevent getting a dupe
                 _keepalive = self._keepalive()
                 loop.create_task(_keepalive)
                 if self._concb is not None:
                     # apps might need to know connection to the server acquired
                     launch(self._concb, True, *self._concbargs)
                 await self._evfail  # Pause until something goes wrong
-                self._verbose and print(self._evfail.value())
                 self._ok = False
                 asyncio.cancel(_reader)
-                asyncio.cancel(_writer)
                 asyncio.cancel(_keepalive)
-                await asyncio.sleep(1)  # wait for cancellation
```

Comment:
execute launch(self._concb) before waiting 1 second so that applications can adapt to the new connection state as quickly as possible.